src/Extension.cpp

```diff
@@ -0,0 +1,337 @@
+    // 'type' element exists.
+    return node["type"];
+  }
+};
+
+template <> struct convert<ScalarFunctionVariant> {
+  static bool decode(const Node &node, ScalarFunctionVariant &function) {
+    return decodeFunctionVariant(node, function);
+  };
+};
+
+template <> struct convert<AggregateFunctionVariant> {
+  static bool decode(const Node &node, AggregateFunctionVariant &function) {
+    const auto &res = decodeFunctionVariant(node, function);
+    if (res) {
+      const auto &intermediate = node["intermediate"];
+      if (intermediate) {
+        function.intermediate = Type::decode(intermediate.as<std::string>());
+      }
+    }
+    return res;
+  }
+};
+
+template <> struct convert<io::substrait::TypeVariant> {
+  static bool decode(const Node &node, io::substrait::TypeVariant &typeAnchor) {
+    auto &name = node["name"];
+    if (name && name.IsScalar()) {
+      typeAnchor.name = name.as<std::string>();
+      return true;
+    }
+    return false;
+  }
+};
+
+} // namespace YAML
+
+namespace io::substrait {
+
+namespace {
+
+std::string getSubstraitExtensionAbsolutePath() {
+  const std::string absolute_path = __FILE__;
+  auto const pos = absolute_path.find_last_of('/');
+  return absolute_path.substr(0, pos) + "/../third_party/substrait/extensions/";
+}
+
+} // namespace
+
+std::shared_ptr<Extension> Extension::load() {
+  static const auto registry = loadDefault();
+  return registry;
+}
+
+std::shared_ptr<Extension> Extension::loadDefault() {
+  static const std::vector<std::string> extensionFiles = {
+      "functions_aggregate_approx.yaml",
+      "functions_aggregate_generic.yaml",
+      "functions_arithmetic.yaml",
+      "functions_arithmetic_decimal.yaml",
+      "functions_boolean.yaml",
+      "functions_comparison.yaml",
+      "functions_datetime.yaml",
+      "functions_logarithmic.yaml",
+      "functions_rounding.yaml",
+      "functions_string.yaml",
+      "functions_set.yaml",
+      "unknown.yaml",
+  };
+  const auto &extensionRootPath = getSubstraitExtensionAbsolutePath();
+  return load(extensionRootPath, extensionFiles);
+}
+
+std::shared_ptr<Extension>
+Extension::load(const std::string &basePath,
+                const std::vector<std::string> &extensionFiles) {
+  std::vector<std::string> yamlExtensionFiles;
+  yamlExtensionFiles.reserve(extensionFiles.size());
+  for (auto &extensionFile : extensionFiles) {
+    auto const pos = basePath.find_last_of('/');
+    const auto &extensionUri = basePath.substr(0, pos) + "/" + extensionFile;
+    yamlExtensionFiles.emplace_back(extensionUri);
+  }
+  return load(yamlExtensionFiles);
+}
+
+std::shared_ptr<Extension>
+Extension::load(const std::vector<std::string> &extensionFiles) {
+  auto registry = std::make_shared<Extension>();
+  for (const auto &extensionUri : extensionFiles) {
+    const auto &node = YAML::LoadFile(extensionUri);
+
+    auto &scalarFunctions = node["scalar_functions"];
```

Comment:
should this be `const`?

Reply:
fixed

Reply:
@chaojun-zhang do you think similar code segments in the following sections would need similar adjustments.